bplustree_test.go

```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
+}
+
+func TestInsertStr(t *testing.T) {
+	t.Log("Starting insert test..")
+	elems := make([]TestingElemStr, numElems)
+	errors := make([]error, numElems)
+	var wg sync.WaitGroup
+
+	for i := 0; i < numElems; i++ {
+		elem := &elems[i]
+		elem.key = makeTestingKeyStr(i)
+		elem.val = i + 10
+		wg.Add(1)
+		go InsertRoutineStr(t, elem, errors, i, &wg)
+	}
+
+	wg.Wait()
+	for i := 0; i < numElems; i++ {
+		if errors[i] != nil {
+			t.Errorf("Failed to insert: %d: %v", i, errors[i])
+			t.FailNow()
+		}
+	}
+	t.Logf("\nFinal tree:\n")
+	newtree.Print()
+}
+
+func TestSearchStr0(t *testing.T) {
+	t.Log("Starting search tests..")
+
+	// case 0. Look for a key which doesn't exist.
+	testingKey := makeTestingKeyStr(10000)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Exact}
+	result, err := newtree.Search(ss)
+	if err == nil {
+		t.Errorf("Expected to not find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result != nil {
+		t.Errorf("Expected to not get any result but got: %v", result)
+		t.FailNow()
+	}
+
+}
+
+func keyEvaluatorStr(k1 Key, k2 Key) bool {
+	result := k1.Compare(k2)
+	if result < 10 && result > -10 {
+		return true
+	}
+
+	return false
+}
+func TestSearchStr1(t *testing.T) {
+	// case 1. Exact search.
+	testingKey := makeTestingKeyStr(100)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Exact}
+	result, err := newtree.Search(ss)
+	if err != nil {
+		t.Errorf("Failed find the key: %v, err: %v", ss.searchKey, err)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Got nil result")
+		t.FailNow()
+	}
+
+	// case 1.a Exact search. Provide an evaluator and maxElems which
+	// should be ignored.
+	testingKey1 := makeTestingKeyStr(100)
+	ss = SearchSpecifier{searchKey: testingKey1, direction: Exact,
+		maxElems: 50, evaluator: keyEvaluatorStr}
+	result, err = newtree.Search(ss)
+	if len(result) > 1 {
+		t.Errorf("Expected to find exact match, but got more (%d)", len(result))
+		t.FailNow()
+	}
+	if err != nil {
+		t.Errorf("Expected to find the key: %v (err: %v)", ss.searchKey, err)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Got nil result")
+		t.FailNow()
+	}
+}
+
+func TestSearchStr2(t *testing.T) {
+	// case 2. maxElems greater than limit.
+	testingKey := makeTestingKeyStr(100)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Exact,
+		maxElems: 100, evaluator: keyEvaluatorStr}
+
+	result, err := newtree.Search(ss)
+	if err == nil {
+		t.Errorf("Expected to not find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result != nil {
+		t.Errorf("Expected to not get any result but got: %v", result)
+		t.FailNow()
+	}
+}
+
+func TestSearchStr3(t *testing.T) {
+	// case 3. maxElems = 10, in Right direction. nil evaluator
+	testingKey := makeTestingKeyStr(100)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Right,
+		maxElems: 10}
+
+	result, err := newtree.Search(ss)
+	if err != nil {
+		t.Errorf("Expected to find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
```

Comment:
all such cases can be replaced by a single line, you can use a package for asserts.